include/zephyr/logging/log.h

```diff
@@ -89,13 +90,189 @@ extern "C" {
  */
 #define LOG_WRN_ONCE(...)					\
 	do {							\
-		static uint8_t __warned;			\
-		if (unlikely(__warned == 0)) {			\
+		static atomic_t __warned = ATOMIC_INIT(0);	\
+		if (unlikely(atomic_cas(&__warned, 0, 1))) {	\
 			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);	\
-			__warned = 1;				\
 		}						\
 	} while (0)
 
+#if defined(CONFIG_LOG_RATELIMIT_ENABLE)
+/**
+ * @brief Writes a WARNING level message to the log with rate limiting.
+ *
+ * @details It's meant for situations that warrant investigation but could clutter
+ * the logs if output too frequently. The message will be logged at most once
+ * per specified interval in milliseconds.
+ *
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_WRN_RATELIMITED_RATE(_rate_ms, ...)					\
+	do {									\
+		static atomic_t __last_log_time = ATOMIC_INIT(0);		\
```

Comment:
statics shouldn't be initialised to `0`, that happens automatically.
Same for all the other macros.